x-pack/plugins/infra/server/lib/alerting/metric_threshold/preview_metric_threshold_alert.ts

```diff
@@ -0,0 +1,170 @@
+) => {
+  // There are three different "intervals" we're dealing with here, so to disambiguate:
+  // - The lookback interval, which is how long of a period of time we want to examine to count
+  //   how many times the alert fired
+  // - The interval in the alert params, which we'll call the bucket interval; this is how large of
+  //   a time bucket the alert uses to evaluate its result
+  // - The alert interval, which is how often the alert fires
+
+  const { timeSize, timeUnit } = params.criteria[0];
+  const bucketInterval = `${timeSize}${timeUnit}`;
+  const bucketIntervalInSeconds = getIntervalInSeconds(bucketInterval);
+
+  const lookbackInterval = `1${lookback}`;
+  const lookbackIntervalInSeconds =
+    overrideLookbackIntervalInSeconds ?? getIntervalInSeconds(lookbackInterval);
+
+  const start = end - lookbackIntervalInSeconds * 1000;
+  const timeframe = { start, end };
+
+  // Get a date histogram using the bucket interval and the lookback interval
+  try {
+    const alertResults = await evaluateAlert(callCluster, params, config, timeframe);
+    const groups = Object.keys(first(alertResults));
+
+    // Now determine how to interpolate this histogram based on the alert interval
+    const alertIntervalInSeconds = getIntervalInSeconds(alertInterval);
+    const alertResultsPerExecution = alertIntervalInSeconds / bucketIntervalInSeconds;
+    const previewResults = await Promise.all(
+      groups.map(async (group) => {
+        const tooManyBuckets = alertResults.some((alertResult) =>
+          isTooManyBucketsPreviewException(alertResult[group])
+        );
+        if (tooManyBuckets) {
+          return TOO_MANY_BUCKETS_PREVIEW_EXCEPTION;
+        }
+
+        const isNoData = alertResults.some((alertResult) => alertResult[group].isNoData);
+        if (isNoData) {
+          return null;
+        }
+        const isError = alertResults.some((alertResult) => alertResult[group].isError);
+        if (isError) {
+          return NaN;
+        }
+
+        // Interpolate the buckets returned by evaluateAlert and return a count of how many of these
+        // buckets would have fired the alert. If the alert interval and bucket interval are the same,
+        // this will be a 1:1 evaluation of the alert results. If these are different, the interpolation
+        // will skip some buckets or read some buckets more than once, depending on the differential
+        const numberOfResultBuckets = first(alertResults)[group].shouldFire.length;
+        const numberOfExecutionBuckets = Math.floor(
+          numberOfResultBuckets / alertResultsPerExecution
+        );
+        return [...Array(numberOfExecutionBuckets)].reduce(
+          (totalFired, _, i) =>
+            totalFired +
+            (alertResults.every(
+              (alertResult) =>
+                alertResult[group].shouldFire[Math.floor(i * alertResultsPerExecution)]
+            )
+              ? 1
+              : 0),
+          0
+        );
+      })
```

Comment:
It could be just me, but this is hard to wrap my head around. It may be easier to read with variables? `const someGoodVariableName = [...Array(numberOfExecutionBuckets)]` `const goodVar2 = (alertResults.every(
              (alertResult) =>
                alertResult[group].shouldFire[Math.floor(i * alertResultsPerExecution)]
            )`